keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1458 @@
+`resources` field is specified at pod-level. 
+
+If the workloads are not using the new feature, downgrade won't be affected. However, if it is required to
+downgrade a cluster that has workloads using the pod-level limits functionality, following steps need to be
+performed:
+
+* Translate pod-level resource specifications to container-level specifications.
+
+* Delete existing workloads and recreate them with container-level specifications.
+
+* Drain the nodes one by one to prevent the new pods from being scheduled on them.
+
+* Downgrade the kubelet on the drained nodes to a version that doesn't support pod-level resource specification
+fields.
+
+* Downgrade the control plane components.
+
+* Uncordon the nodes, making it available for scheduling again.
+
+
+### Version Skew Strategy
+
+Pod-level resource specification is an opt-in feature. For this feature to work correctly, it must be enabled
+in all parts of the cluster (scheduler, API server, kubelet). However, if a node is running an older version
+of kubelet that doesn't support this feature, it will ignore the pod-level resouce specifications and continue
+to use the aggregated container-level values (backward compatible). Applications must be prepared for this version
+skew implementation.
+
+For users wanting to use this feature, it is user's responsibility to use the correct version of kube-scheduler,
+kube-apiserver and kubelet otherwise the feature will not work as expected.
+
+## Production Readiness Review Questionnaire
+
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: PodLevelResources
+  - Components depending on the feature gate: kubelet, kube-apiserver, kube-scheduler
+  - Will enabling / disabling the feature require downtime of the control
+    plane? No. `resources` must be set/unset in `spec` at pod-level to enable/disable this feature.
+    The workloads need to restart to disable the feature.
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? No. Only the workloads need to restart.
+
+###### Does enabling the feature change any default behavior?
+
+No. This feature is guarded by a feature gate, and requires setting Pod level `resource` stanza explicitly.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
```

Comment:
one fun aspect - if pods with pod limits were scheduled while feature gate was enabled, they will be scheduled on nodes with no resources.

Also CA may go highwire: https://github.com/kubernetes/enhancements/pull/4678/files#r1645206055

Reply:
@SergeyKanzhelev Can you explain more? I'm not sure I follow, but want to make sure this is resolved before PRR approval.

Reply:
Does this refer to pods with pod level resource requests (but no container level requests) may get scheduled onto nodes without sufficient resources when the feature gate is **disabled**? This seems to be an edge case that would only happen if the feature gets disabled AFTER the pod creation but BEFORE the pod scheduling.

Reply:
I think beyond the edge case, since CA doesn't take pod-level resources into account, if a user specifies only pod-level resources then CA will think that this pod doesn't need any resources as container-level values are not set which could match such pods to nodes with no resources. I'd say this is expected for now, until the support for CA comes in with Beta

Reply:
yes, I meant that if somebody scheduled a bunch of deployments with pod level resources, and then feature was disabled, those Pods will be seen as not having any resources declared. And may cause some troubles in how they will be scheduled